src/OpenAIGym.jl

```diff
@@ -19,40 +20,47 @@ const _py_envs = Dict{String,Any}()
 abstract type AbstractGymEnv <: AbstractEnvironment end
 
 "A simple wrapper around the OpenAI gym environments to add to the Reinforce framework"
-mutable struct GymEnv <: AbstractGymEnv
+mutable struct GymEnv{T} <: AbstractGymEnv
     name::String
-    pyenv  # the python "env" object
-    state
+    pyenv::PyObject   # the python "env" object
+    pystep::PyObject  # the python env.step function
+    pyreset::PyObject # the python env.reset function
+    pystate::PyObject # the state array object referenced by the PyArray state.o
+    pystepres::PyObject # used to make stepping the env slightly more efficient
+    info::PyObject    # store it as a PyObject for speed, since often unused
+    state::T
     reward::Float64
     total_reward::Float64
     actions::AbstractSet
     done::Bool
-    info::Dict
-    GymEnv(name,pyenv) = new(name,pyenv)
-end
-GymEnv(name) = gym(name)
-
-function Reinforce.reset!(env::GymEnv)
-    env.state = env.pyenv[:reset]()
-    env.reward = 0.0
-    env.total_reward = 0.0
-    env.actions = actions(env, nothing)
-    env.done = false
+    function GymEnv{T}(name, pyenv, pystate, state) where T
+        env = new{T}(name, pyenv, pyenv["step"], pyenv["reset"],
+                                 pystate, PyNULL(), PyNULL(), state)
+        reset!(env)
+        env
+    end
 end
 
-function gym(name::AbstractString)
+function GymEnv(name; stateT=PyArray)
```

Comment:
How about this API design?
```julia
GymEnv(name::Symbol, ver::Symbol = :v0; ...)

julia> GymEnv(:CartPole)  # implies "CartPole-v0" in Python
julia> GymEnv(:CartPole, :v1)
```

Reply:
btw, the env name is a constant, so I think we can do something more aggressively,
```julia
mutable struct GymEnv{N,V,T}
```
e.g.`GymEnv{:CartPole, :v1, ...}`
Than we can remove the branch in this via multiple dispatch.

Reply:
I'm trying to figure out why we need `stateT`

Reply:
> GymEnv(:CartPole, :v1)
Yeah I quite like that

> Than we can remove the branch in this via multiple dispatch.
Not sure if people tend to create envs in tight loops, so may not be that important to speed up the constructor. 

> I'm trying to figure out why we need stateT
Blackjack-v0 isn't a numpy array, i.e. can't be converted to PyArray, you might also want to auto-convert to say regular Array

Reply:
> Not sure if people tend to create envs in tight loops, so may not be that important to speed up the constructor. Though may be there could be some other use cases that might benefit?

I do not care speed in this case, I just do not like `if` branch :p

Reply:
Lol, what about: https://docs.julialang.org/en/stable/manual/performance-tips/#The-dangers-of-abusing-multiple-dispatch-(aka,-more-on-types-with-values-as-parameters)-1 ?

Reply:
:trollface: 

Acctually, I think we need something different for `GymEnv{:Blackjack,:v0}`.
But I don't fully figure out it yet.

Reply:
I'm reading the Blackjack source code.
It's quite simple.
https://github.com/openai/gym/blob/master/gym/envs/toy_text/blackjack.py

If we add the type parameters `GymEnv{:Blackjack, ...}`, 
I can port it in pure Julia, so we can remove all `stateT` stuffs.

Reply:
Well someone still might want to use a different state type, e.g. Array.

Reply:
ah, I see.

Reply:
But the next question:
I think user will might want to issue `GymEnv("Blackjack-v0")` intuitively, not `GymEnv("Blackjack-v0", stateT = PyAny)`.

If Blackjack is the only case, parametric type can solve this.
If there are more cases beyond our imagination, we will need something for auto-detection.

Reply:
Yeah, this is true, I was planning to put a note in the README, but that's not ideal. 

We could just put `GymEnv(envname, stateT = envname=="Blackjack-v0"? PyArray: PyAny)` ?

Reply:
There can be more cases, but maybe we just wait for them to be filed as issues?

Reply:
> We could just put GymEnv(envname, stateT = envname=="Blackjack-v0"? PyArray: PyAny) ?

Yes.
I still want parametric type for it (in long term), but in this PR, it's fine already.